### ISSUE TITLE
Support for ember-template-imports?

### ISSUE THREAD

**Issue by deanylev** [salsify/ember-css-modules#318]
Are there any plans to support [ember-template-imports](https://github.com/ember-template-imports/ember-template-imports)? We were planning to switch our new components over to that but unfortunately it appears ember-css-modules is incompatible. The outputted HTML simply includes the `local-class` attribute verbatim, instead of transforming it to the mangled `class` attribute.

**Comment by deanylev** [comment]
Appears the `local-class` helper works in the meantime which leads me to think fixing this properly shouldn't be too hard. I've taken a look myself but I'm not really sure why this local class injection hack works in normal components but not `gjs`/`gts` ones

**Comment by deanylev** [comment]
Narrowed it down to this line https://github.com/salsify/ember-css-modules/blob/409cc26b93cca9dadd14c7825fb393abfd1be182/packages/ember-css-modules/lib/htmlbars-plugin/index.js#L52 but doesn't work due to strict mode 😕 wonder if the automagic `local-class` attribute replacement will be possible in that case

**Comment by dfreeman** [comment]
We're planning on a pretty major overhaul (or really, stripping down) of `ember-css-modules`. The end result will lean heavily on the bundler (Webpack via `ember-auto-import`, or Webpack/Vite via Embroider) to handle the actual "CSS Modules" of it all, and the only custom bit left will be the `local-class` transform.

The goal there will be full compatibility with Embroider's strictest settings and FCCTs (a.k.a `ember-template-imports`) for those who want to keep using the `local-class` attribute going forward rather than importing and directly referencing the `styles` hash themselves.

To temper your expectations, though, I currently have that work lined up to likely happen around May or June, depending on how a couple of other internal projects I'll be taking on at work over the next few months go.

**Comment by deanylev** [comment]
@dfreeman any update on when support for this will be added? 🙏

**Comment by dfreeman** [comment]
Yes! I started on this work in June, but have been on vacation for the past six weeks—I start back at work on Monday, and will keep you updated here as there's progress :)

**Comment by ajitsinghkamal** [comment]
Hey, I haven't dived deep into the code yet and new to ember as well but first of all, thank @dfreeman for this plugin. I am not sure if @deanylev found a solution but I figured out a hacky way to use module classes work in `.gts|.gjs` for now. 

It seems like importing the stylesheet and accessing the class through the import works for most cases.(skip the scss extension)

```css
.shell {
  display: grid;
  grid-template-rows: var(--app-header-ht) 1fr;
  grid-template-columns: 1fr;
  height: 100%;
  width: 100%;
  overflow: hidden;
}
```

```ts
// @ts-expect-error
import css from './styles';
import { get } from '@ember/object';

const Shell = <template>
  <div class='{{get css "shell"}}' ...attributes>
    <Header />
    <main class='page-container'>
      {{yield}}
    </main>
  </div>
</template>;
```
PS. I too think this would be a great improvement so happy to contribute if needed :)

**Comment by RobbieTheWagner** [comment]
@dfreeman any updates on this refactor?

**Comment by dfreeman** [comment]
@RobbieTheWagner I'm still working on writing up documentation, but I was able to wrap up the implementation work on that refactor during a company hackathon recently. That work lives on the [`standalone-template-transform` branch](https://github.com/salsify/ember-css-modules/tree/standalone-template-transform) and has been published as `ember-css-modules@3.0.0-alpha.1`.

The v3 alpha should work well with FCCTs (`.gjs`/`.gts` files) as well as `@embroider/webpack` with its various `static*` flags enabled; see [the `app-v1-ecm` test package](https://github.com/salsify/ember-css-modules/tree/standalone-template-transform/test-packages/app-v1-ecm).

That branch also introduces a `glimmer-local-class-transform` package that can be used in v2 apps and addons, as well as an `ember-local-class` package for v1 apps that _only_ installs the `local-class` syntax for templates and allows you to rely on your bundler for actual CSS Modules support rather than Broccoli shenanigans. See [the various `test-packages`](https://github.com/salsify/ember-css-modules/tree/standalone-template-transform/test-packages) for examples.

All three packages are available as alpha release if you'd like to test drive them. As I said I'm still working on real documentation about how the packages fit together and migration details, but here's the outline I'm working from in case it's helpful:

```
Breaking changes:
 - Min Node version 22
 - Min Ember version is 3.28; also requires function helpers polyfill if < 4.5
 - Removed the plugin system
 - Template association is now based on a regex map (should work essentially as before, but could be edge cases)

Differences between `ember-css-modules` and `ember-local-class`:
  - ELC defaults to `.module.css` rather than any `.css`
  - ELC defaults to only supporting colocated CSS (no classic/pods layout)
  - ELC does not do any CSS preprocessing; it just sets up `glimmer-local-class-transform` (which provides support for the `local-class` attr and helper)

v1 App (pick one):
 - `ember-css-modules` (classic build or `@embroider/webpack`) OR
 - `ember-local-class` + `ember-auto-import` with `allowAppImports` and `modules: { auto: true }` OR
 - `ember-local-class` + `@embroider/webpack` with `modules: { auto: true }`

v1 Addon:
 - `ember-css-modules` in `dependencies`

v2 App:
 - `glimmer-local-class-transform` in `babel.config.cjs`

v2 Addon:
 - `glimmer-local-class-transform` in `babel.publish.config.cjs`
 - `.hbs` in Babel extensions and `keepAssets(..., 'default')` in `rollup.config.mjs`
 - Either set up your own CSS Modules preprocessing or require host to deal with it (the latter is better for treeshaking and gives the host more say in how CSS is bundled)
```

**Comment by musaffa** [comment]
@dfreeman Any update on ember-css-modules 3.0?

**Comment by musaffa** [comment]
Since ember 6.8, Vite build system has become the default. This addon has stopped working.

@dfreeman Can you please look into it?

**Comment by deanmarano** [comment]
@dfreeman I've been testing ember-css-modules@3.0.0-alpha.1 with a 6.4 broccoli-based app and ran into a runtime error where 

`classNames throws TypeError: Cannot read properties of undefined (reading 'trim') when undefined is passed as the classes argument.` 

Additionally, `classNames()` in the runtime was passing through unmapped class names (via `??` fallback) instead of dropping them. This meant our global utility classes (like `.is-hidden`) placed inside local-class were being applied to the DOM, which wasn't the case in v2. This broke CSS transitions in our app because `display: none !important` was being applied immediately, preventing `transitionend` events.

I have fixes with tests here: https://github.com/salsify/ember-css-modules/pull/335

This allows our test suite (~6k tests) to pass without any changes.

I'd be happy to help document the upgrade path — the outline in the comment above is a great start and I could help flesh it out into proper docs.

**Comment by dfreeman** [comment]
Thank you @deanmarano! I promise I meant well for what that's worth, but clearly I haven't been able to devote the time that I would've liked to this 😅 

I'd love to have your help getting things documented and cleaned up for release!

I've pushed up the skeletal docs for `ember-local-class` and `glimmer-local-class-transform` that were sitting in my working copy, and I'll aim to carve out some time tomorrow to figure out what's breaking under Windows on #335 (clearly unrelated to your change) so that that's unblocked.

**Comment by RobbieTheWagner** [comment]
@dfreeman if you would like to add us as maintainers, we would be happy to work on pushing this forward 😃

**Comment by dfreeman** [comment]
@RobbieTheWagner Unfortunately adding external maintainers is complicated for Salsify-owned repos due to some process requirements we have.

**Comment by deanmarano** [comment]
I've opened #338 which adds migration guides for both apps and addons, along with fleshed-out READMEs for glimmer-local-class-transform and rollup-plugin-preprocess-css-modules.

The migration guides cover the path from classic Broccoli builds through Embroider+Webpack to Embroider+Vite, so they should help folks looking to adopt the new packages that support .gjs/.gts files.

Would love any feedback — especially from anyone who's tried the alpha or is looking to migrate!

**Comment by dfreeman** [comment]
#340 has been merged and released! Please give the new packages/versions and don't hesitate to reach out with issues.

Thanks again to @deanmarano for pushing this over the finish line 🙏
